02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md

```diff
@@ -0,0 +1,196 @@
+# State of PureScript Documentation: 2019
+
+This document is not meant to convince you to use PureScript, we assume that readers are already interested in it or using it.
+
+Numerous parties within the PureScript community (e.g. [core contributors](), [users](), [new learners]()) believe that the documentation is lacking in some critical areas. As a result, @chexxor started the [PureScript documentation efforts in 2019](https://discourse.purescript.org/t/purescript-documentation-efforts-in-2019/524) Discourse thread to answer one question:
+> How can we improve the PureScript documentation in 2019?
+
+(Purpose) The following document tries to do the following things:
+1. Provide a researched narrative of PureScript's documentation development and history.
```

Comment:
This point isn't addressed by looking at the ToC. I'll spend time in the next few days to understand this difference and find a way to resolve it.

Reply:
A related thing I noticed: The first thing that is discussed in this document is "What is Good Documentation?", from the ToC, but I expect to learn about the "State of PureScript Documentation", as that's the title of the document.

Reply:
The purpose of this step, from its ReadMe:

> A "context" or "narrative" is a person or group of people's attempt at integrating their isolated interpretations of various sources into a coherent "story" that explains all the relevant pieces of the puzzle in a systematic way.

So, we have accumulated "interpretations of various sources" from a previous step, and it sounds like this step is to integrate those into a single "story". To explain the "explains all the relevant pieces" part, then: When putting a piece into this "story", it should be self-evident that it does indeed belong in the story.

Reply:
It's a little difficult to create a "story" when given no "grounding" or constraints, so let's say choose the grounding to be "How can we improve the PureScript documentation in 2019", as it was the original motivation question of this project.

1. How can we improve the PureScript documentation? This question is doesn't have a clear answer, so we need to break it down into smaller questions:
2. What *should* documentation look like?
3. How does *current* PureScript documentation compare?
4. *Why* is PureScript's documentation lacking in the places it is?
5. How should we improve these weak areas of documentation?

Judging purely from the current ToC, it's not clear where the "Addressing Audiences" section belongs in the above list of questions.

(I realize I've discussed the ToC issue with @JordanMartinez before, but this current ToC still bothers me, perhaps because I don't see how it's relevant to the title of the document and the overall goal of the project.)

Reply:
In https://github.com/chexxor/purescript-documentation-discussion/pull/50#issuecomment-468945658, Jordan says:

> Outcome
> * learners know how to best learn in this situation
> * doc writers know guidelines to follow, have a 'checklist' to ensure they have 'good' documentation in this regard, and have 'templates' they can use to quickly get started
> * core contributors.... Well, I'm not sure what the outcome should be here
> * users of PS... same as core contributors: what outcome should be here?

It looks like Jordan implicitly broke down the "How can we improve the PureScript documentation?" question into those four topics. This is different than how I broke down that question just now in my previous comment. I think this is a primary reason for the difference in ToCs we are preferring.

Reply:
Perhaps we can unify our two ways of breaking down the "How can we improve the PureScript documentation?" question.

1. How can we improve the PureScript documentation? "Documentation" targets a specific audience by nature, so we need to break this question down per audience:
2. Documentation for PureScript Learners.
    1. What should documentation look like?
    2. How does current PureScript documentation compare?
    3. Why is this documentation lacking in the places it is?
    4. How should we improve these weak areas of documentation?
3. Documentation for Documentation Writers.
    1. What should documentation look like?
    2. How does current PureScript documentation compare?
    3. Why is this documentation lacking in the places it is?
    4. How should we improve these weak areas of documentation?
4. Documentation for Core Contributors.
    1. What should documentation look like?
    2. How does current PureScript documentation compare?
    3. Why is this documentation lacking in the places it is?
    4. How should we improve these weak areas of documentation?
5. Documentation for Users of PS.
    1. What should documentation look like?
    2. How does current PureScript documentation compare?
    3. Why is this documentation lacking in the places it is?
    4. How should we improve these weak areas of documentation?

It's quite a clear ToC.

Whichever ToC we choose here, we'll have to update the content of this document to fit into this new structure. That can be a separate project, I think.

Reply:
I'm exploring this different structure here: https://github.com/chexxor/purescript-documentation-discussion/pull/67

Reply:
> This point isn't addressed by looking at the ToC. I'll spend time in the next few days to understand this difference and find a way to resolve it.

Is the issue here that the ToC doesn't seem to show how this point is explained? I'm not sure how to understand this comment.

> The first thing that is discussed in this document is "What is Good Documentation?", from the ToC, but I expect to learn about the "State of PureScript Documentation", as that's the title of the document.

If we changed that header to "Defining a Criteria: What is Good Documentation?" would that help address this point you raise?

> So, we have accumulated "interpretations of various sources" from a previous step, and it sounds like this step is to integrate those into a single "story". To explain the "explains all the relevant pieces" part, then: When putting a piece into this "story", it should be self-evident that it does indeed belong in the story.

I'm reading the above as "Alex is thinking out loud. He doesn't expect me to respond to it."

> It looks like Jordan implicitly broke down the "How can we improve the PureScript documentation?" question into those four topics. This is different than how I broke down that question just now in my previous comment. I think this is a primary reason for the difference in ToCs we are preferring.

Mm..... not quite. Look at the 'Why read this document?' header and the text that later appears towards the bottom of this document (search for "A few different audiences") in [one of the first commits in this PR](https://github.com/chexxor/purescript-documentation-discussion/blob/3f8dee84f77264b00d53b4c929d496e93bef29c8/02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md#why-read-this-document). The outline's first three parts were the heart of the document. If we deleted the "various audiences" part, the document would still stand. However, I thought it'd be worthwhile to say something directly to specific audiences now that they have, presumably, read the heart of the document. 

I'd like to address another aspect of your comment that touches on your idea for how to structure the ToC in #67. Yes, the content I wrote in addressing the "new learners" audience and "PS doc writers" audience are specific to documentation: one reads what another writes. However, the "core contributors" audience isn't documentation-related per-say. Rather, it asks how we can support them and asks them to clarify their position on some things (e.g. whether they are willing to pass maintainership to others or not and under what criteria). Another idea I had was to address yet another audience (PS Users), asking them what libraries would make the ecosystem more mature. This question isn't directly documentation-related, but it does affect the overall ecosystem of things.

Thus, when I see your ToC structure, some things don't make sense to me:
- Why should PS learners care what 'docs should look like?' Isn't that more of a Doc Writers concern?
- Why should core contributors care about what 'docs should look like' or how 'ps docs compare' ? Do all of them write docs? A few do, but then I'd address them as doc writers, not core contributors. One person can fulfill both roles but I'll address such a person in the role that is relevant to the work that he/she is doing.
- How would we explain why docs are lacking in 3 different places? Moreover, why would we do that? It seems to me that the only way to do it well would be to duplicate the same content in 3 places.

The main heart of the document (as it presently stands) is still the first three sections:
- [What is Good Documentation anyway?](https://github.com/chexxor/purescript-documentation-discussion/blob/6bafbe6a140e01fc74c7cbfec879f508e136de7a/02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md#what-is-good-documentation-anyway) - the criteria
- [Evaluating PS Docs](https://github.com/chexxor/purescript-documentation-discussion/blob/6bafbe6a140e01fc74c7cbfec879f508e136de7a/02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md#evaluating-purescripts-documentation) - the evaluation
- [Why is PS docs lacking?](https://github.com/chexxor/purescript-documentation-discussion/blob/6bafbe6a140e01fc74c7cbfec879f508e136de7a/02-Context-or-Narrative/State-of-PureScript-Documentation-2019.md#why-is-purescripts-documentation-lacking) - the explanation behind the evaluation

Reply:
> Is the issue here that the ToC doesn't seem to show how this point is explained? I'm not sure how to understand this comment.

Yeah, when I read the document, the `1. Provide a researched narrative of PureScript's documentation development and history.` goal isn't fulfilled. I would expect a narrative of the history of the development and growth of the PureScript project and ecosystem, but that's quite an ambitious goal and we actually haven't even spent time finding sources for that kind of goal. IIRC, I added that `1. Provide a researched...` point and now I think it was an error on my part. I think it's best to just remove that goal from here.

> If we changed that header to "Defining a Criteria: What is Good Documentation?" would that help address this point you raise?

Yeah, I think that would improve the clarity a lot.

You ask why should PS learners or core contributors care about what docs should look like. These groups work with PureScript, and I think anyone who works in the ecosystem will need documentation of some kind. For example, a core contributor might want documentation on the type theory implemented in the compiler, or a well-defined set of principles written down to reference when making decisions of merging pull requests, or a language specification to reference when understanding the finer details of an unfamiliar part of the language.

Likewise for PS learners - they care about what docs should look like because they'll be consuming them. It's important to define the needs and expectations of each audience to reference when creating documentation targeting that audience.

> Look at the 'Why read this document?' header and the text that later appears towards the bottom of this document (search for "A few different audiences") in one of the first commits in this PR.

I have yet to read and digest the stuff referenced in this part of your comment. I intend to post another comment about that later. I expect this to be an important realization for me, as our two approaches here seem slightly different in an important way.

Reply:
> Yeah, I think that would improve the clarity a lot.

Done!

> You ask why should PS learners or core contributors care about what docs should look like. These groups work with PureScript, and I think anyone who works in the ecosystem will need documentation of some kind. For example, a core contributor might want documentation on the type theory implemented in the compiler, or a well-defined set of principles written down to reference when making decisions of merging pull requests, or a language specification to reference when understanding the finer details of an unfamiliar part of the language.

Just a thought, but it seems like you're using the term `documentation` to refer to content that targets new learners (which is what I thought we were doing) and content that targets contributors to the PureScript language itself (which I assumed was outside this project's scope). While I'm sure the latter could benefit from that (or benefit from defining specific procedures to follow for various activities), I don't think that's what we're trying to do here.

> our two approaches here seem slightly different in an important way.

:smiley: I think it's more than just "slightly" :wink:

Reply:
I use the term `documentation` to refer to any written material that answers questions about a project. To me, the primary value of documentation is that I don't need to be in the rare state of having a person who is knowledgeable about the project and also available to answer my questions **and also** able to clearly communicate the answers to me. Core contributors and documentation writers both have questions about a project that need answering.

Regarding the ToC/structure of the document, when I was first trying to understand the structure of the document I think I viewed the "addressing various audiences" part as an essential part of the document. On my own, I couldn't think of a way to reasonably have that as a primary part of this document while keeping its scope/goal/content as it was, so I started to get creative with it. It seems now that I may have misunderstood that part's importance. If it's not an essential part of this document, then perhaps we can move it somewhere else? The following paragraph explains my thinking on this proposal.

I intended this document's authors and audience to be documentation managers, to be a unified understanding of the situation. Based on this document, then, documentation managers could define "outcomes" and clear steps towards them. Some guesses I have about our misunderstanding: 1) It seems you want this document's audience to be contributors and even consumers of PureScript's documentation, or perhaps 2) You intended to include some Outcomes and steps towards them in this narrative. I might disagree with having these two things as part of this document.

What do you think? Am I getting closer to the core of our misunderstanding?

In the meantime, I'll review this document again, in light of my new understanding of your original intentions (what I think were original).

Reply:
Ah, I forgot that we had talked about it maybe making sense for this document to contain desired outcomes. I'll have to find that past discussion.

Perhaps we just need to have a ToC header for that section which is more descriptive of it's relation to the other stuff in this document.

Reply:
> I use the term documentation to refer to any written material that answers questions about a project. To me, the primary value of documentation is that I don't need to be in the rare state of having a person who is knowledgeable about the project and also available to answer my questions and also able to clearly communicate the answers to me.

Ah... Then perhaps we're disagreeing about the 'scope' of the projects in question? I like your definition of `documentation` and I agree with it. However, I have not been including the PureScript language repository as a project that we should include when discussing PS' docs, mainly because most people who want to learn PureScript aren't going to immediately contribute to that.

Perhaps our disagreement lies in two aspects: using the same terms to refer to different meanings, having emphasis on different aspects of the work, and having different breadths of scopes.

For example, I intended the audience of this work to be all of the following--curious outsiders who are interested in the language and keeping updated about it, but who haven't started using it; new learners; people who already know and are familiar with it; and core contributors (primarily language developers, but people who contribute to build tools and whatnot as well), so they can stay in the loop of the conversation and add their input--as a way to explain "Here's where we are in things and why we are there (so we can have some parameters/guidelines to help us know what the problem is and how to fix it). Now that we know more about the problem, what can we do to fix it (answering Alex's original question)?"
However, my emphasis was largely on new learners in general (perhaps because I picked up that 'bias' when reading through our sources). I assumed we had a few goals in mind:
- improving the new learner onboarding experience
- improving Pursuit docs throughout the ecosystem
- explaining more FP ideas and best practices in general, so that people can actually be productive with the language

I did not include these goals:
- help people know how to contribute to the PureScript language
- help core contributors figure out whether and how to improve their administrative policies/procedures (something I'm realizing right now as I type this)

> Some guesses I have about our misunderstanding: .... You intended to include some Outcomes and steps towards them in this narrative. I might disagree with having these two things as part of this document.

Originally, I did not want the Outcome parts included here. Over time, as I continued working on this document, Outcomes started to seep in as my mindset thought of ideas for solving various problems. After a while, I thought it might be best to include them here (via the Outcomes file link) since people would also be in similar mindsets and might come up with good ideas.

Reply:
A side-note: I think our misunderstanding of how to develop this document will serve as great reference when considering ways to prevent this from occurring in the future, because I imagine this problem is an essential problem of collaborative documentation. It seems important to define the primary terms used in the document and to clearly define the goal & audience and perhaps the non-goals and non-audience, and more items as they are discovered.

Regarding the audience of this document, I think we can certainly narrow the scope to only the "new learners" audience. However, I think it's important to acknowledge the existence of the other audiences and to explain why they are not explored. (I imagine they will be explored at a future time, perhaps by a different doc team.) I think this is important to do because when the non-goals are explicit, the boundary/scope of the project is more clear and serves as a "tangible boundary" to reference when reviewing a documentation contribution.

Here's a patch which nearly completely resolves my concerns with this document - https://github.com/chexxor/purescript-documentation-discussion/pull/68 I will think about the "Addressing Various Audiences" section next.

Regarding the Outcomes exploration in this document - I think we can include that in this document. However, as they are right now I don't clearly read them as Outcomes. I'll consider the wording I would expect to see to read it as Outcomes.

Reply:
Yeah, things were going well in terms of our collaboration up until this Context/Narrative part.

Ok, I'll wait to hear from you regarding the Outcomes stuff.